apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/entity/Item.java

```diff
@@ -28,7 +28,7 @@
 
 @Entity
 @Table(name = "Item")
-@SQLDelete(sql = "Update Item set isDeleted = 1 where id = ?")
+@SQLDelete(sql = "Update Item set IsDeleted = 1, DeletedAt = UNIX_TIMESTAMP() where Id = ?")
```

Comment:
UNIX_TIMESTAMP() is the seconds since '1970-01-01 00:00:00' UTC which I think is enough for most cases. However I have some concern that there might be cases it's not enough, e.g. some user uses open api to create item, delete item and create it again that 2 records with the same namespaceId, key and deletedAt exist in the database.
Maybe we could consider using milliseconds? e.g.

```suggestion
@SQLDelete(sql = "Update Item set IsDeleted = 1, DeletedAt = ROUND(UNIX_TIMESTAMP(NOW(4))*1000) where Id = ?")
```
This involves a complex statement but since it's a one time job, so it should be ok.

Reply:
Ignore the implementation, `milliseconds` is indeed better. As you said it's a one-time job, I think it is okay.

By the way, the CI run failed because the H2 doesn't support the MySQL functions, do you have any good ideas?
Some articles said registering the function to H2, but I am not sure it is a good way.

Reply:
Something ran wrong with the function `ROUND(UNIX_TIMESTAMP(NOW(4))*1000)`.

```
Caused by: java.lang.IllegalArgumentException: org.hibernate.QueryException: function takes no arguments: now [UPDATE com.ctrip.framework.apollo.common.entity.AppNamespace SET IsDeleted = 1, DeletedAt = ROUND(UNIX_TIMESTAMP(NOW(4))*1000), DataChange_LastModifiedBy = ?3 WHERE AppId=?1 and Name = ?2]
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:138)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:188)
	at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:725)
	at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:113)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.orm.jpa.ExtendedEntityManagerCreator$ExtendedEntityManagerInvocationHandler.invoke(ExtendedEntityManagerCreator.java:362)
	at com.sun.proxy.$Proxy164.createQuery(Unknown Source)
	at org.springframework.data.jpa.repository.query.SimpleJpaQuery.validateQuery(SimpleJpaQuery.java:87)
	... 59 common frames omitted
Caused by: org.hibernate.QueryException: function takes no arguments: now [UPDATE com.ctrip.framework.apollo.common.entity.AppNamespace SET IsDeleted = 1, DeletedAt = ROUND(UNIX_TIMESTAMP(NOW(4))*1000), DataChange_LastModifiedBy = ?3 WHERE AppId=?1 and Name = ?2]
	at org.hibernate.QueryException.generateQueryException(QueryException.java:120)
	at org.hibernate.QueryException.wrapWithQueryString(QueryException.java:103)
	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.doCompile(QueryTranslatorImpl.java:220)
	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.compile(QueryTranslatorImpl.java:144)
	at org.hibernate.engine.query.spi.HQLQueryPlan.<init>(HQLQueryPlan.java:113)
	at org.hibernate.engine.query.spi.HQLQueryPlan.<init>(HQLQueryPlan.java:73)
	at org.hibernate.engine.query.spi.QueryPlanCache.getHQLQueryPlan(QueryPlanCache.java:162)
	at org.hibernate.internal.AbstractSharedSessionContract.getQueryPlan(AbstractSharedSessionContract.java:604)
	at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:716)
	... 67 common frames omitted
Caused by: org.hibernate.QueryException: function takes no arguments: now
	at org.hibernate.dialect.function.StaticPrecisionFspTimestampFunction.render(StaticPrecisionFspTimestampFunction.java:56)
	at org.hibernate.hql.internal.ast.SqlGenerator.endFunctionTemplate(SqlGenerator.java:226)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.methodCall(SqlGeneratorBase.java:2583)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.simpleExpr(SqlGeneratorBase.java:3102)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.expr(SqlGeneratorBase.java:1645)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.arguments(SqlGeneratorBase.java:4580)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.methodCall(SqlGeneratorBase.java:2554)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.simpleExpr(SqlGeneratorBase.java:3102)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.expr(SqlGeneratorBase.java:1645)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.nestedExpr(SqlGeneratorBase.java:4556)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.multiplicativeExpr(SqlGeneratorBase.java:4233)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.arithmeticExpr(SqlGeneratorBase.java:2799)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.simpleExpr(SqlGeneratorBase.java:3128)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.expr(SqlGeneratorBase.java:1645)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.arguments(SqlGeneratorBase.java:4580)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.methodCall(SqlGeneratorBase.java:2554)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.simpleExpr(SqlGeneratorBase.java:3102)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.expr(SqlGeneratorBase.java:1645)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.binaryComparisonExpression(SqlGeneratorBase.java:3384)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.comparisonExpr(SqlGeneratorBase.java:1453)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.setClause(SqlGeneratorBase.java:1386)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.updateStatement(SqlGeneratorBase.java:354)
	at org.hibernate.hql.internal.antlr.SqlGeneratorBase.statement(SqlGeneratorBase.java:156)
	at org.hibernate.hql.internal.ast.exec.SimpleUpdateExecutor.<init>(SimpleUpdateExecutor.java:56)
	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.buildAppropriateStatementExecutor(QueryTranslatorImpl.java:616)
	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.doCompile(QueryTranslatorImpl.java:208)
	... 73 common frames omitted
```


> e.g. some user uses open api to create item, delete item and create it again that 2 records with the same namespaceId, key and deletedAt exist in the database.

Does this case really need to be considered? Or we just Keep It Simple And Stupid.

Reply:
> Something ran wrong with the function ROUND(UNIX_TIMESTAMP(NOW(4))*1000).

I tried in my local environment and seems it worked well.. 

> By the way, the CI run failed because the H2 doesn't support the MySQL functions, do you have any good ideas?
> Some articles said registering the function to H2, but I am not sure it is a good way.

This is something we need to solve, does it work to register the function to h2?

Reply:
> I tried in my local environment and seems it worked well..

I am sure that the NOW(4) ONLY works with the `@SQLDelete` but not with the `@Modifying @Query`.

As the @query needs to recognize the now() function with arguments, I extend a custom function with Hibernate.
```yaml
spring:
  jpa:
    properties:
      hibernate:
        metadata_builder_contributor: com.ctrip.framework.apollo.common.jpa.SqlFunctionsMetadataBuilderContributor
```

Reply:
OK, that explains. As I only tested the `@SQLDelete` scenario.
How about the CI? Is there a solution to make it work with H2?

Reply:
> How about the CI? Is there a solution to make it work with H2?

I have to say that adapting the H2 is complicated. I have fixed the `NOW` function, but I need to solve another problem, the `UNIX_TIMESTAMP` function.  Maybe I also need to solve the `ROUND` function.

```
Caused by: org.h2.jdbc.JdbcSQLException: Function "UNIX_TIMESTAMP" not found; SQL statement:
Update GrayReleaseRule set IsDeleted = 1, DeletedAt = ROUND(UNIX_TIMESTAMP(NOW(4))*1000) where Id = ? [90022-191]
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
        at org.h2.message.DbException.get(DbException.java:179)
        at org.h2.message.DbException.get(DbException.java:155)
        at org.h2.command.Parser.readJavaFunction(Parser.java:2405)
        at org.h2.command.Parser.readFunction(Parser.java:2457)
        at org.h2.command.Parser.readTerm(Parser.java:2791)
        ...
```

According to https://h2database.com/html/features.html#user_defined_functions, we could define such a function:

```
CREATE ALIAS IF NOT EXISTS UNIX_TIMESTAMP FOR "com.ctrip.framework.apollo.common.jpa.H2Function.unixTimestamp";
```
But it still has a lot of adaptation problems, is this the right way? Maybe we should consider use container MySQL as the Apollo test DataSource, with container like Docker.